File: core/nli-classifier.js
/**
 * Agent Content Shield — NLI Intent Classifier (Layer 2.5)
 *
 * Replaces embedding-similarity with Natural Language Inference.
 * Instead of "does this text LOOK LIKE known injection?"
 * asks "does this text INTEND to do something malicious?"
 *
 * This is vocabulary-independent: passive voice, synonym chains,
 * legal framing, Socratic questions — all caught because the
 * classifier reasons about INTENT, not WORDS.
 *
 * Uses Claude Haiku for speed (~200ms) with Sonnet fallback for
 * borderline cases. Falls back to local deepseek-r1:8b if no API key.
 *
 * Architecture:
 *   Layer 1: Regex (instant, pattern matching)
 *   Layer 2: Embedding scan (50ms, catches near-matches to seeds)
 *   Layer 2.5: NLI Intent Classifier (200ms, catches paraphrased intent) ← NEW
 *   Layer 3: Full LLM analysis (500ms, borderline tiebreaker)
 */

// ═══════════════════════════════════════════════════════════════════════
// THREAT INTENTS — What the attacker is TRYING TO DO, not how they say it
// ═══════════════════════════════════════════════════════════════════════

const fs = require('fs');
const path = require('path');

// ═══════════════════════════════════════════════════════════════════════
// SHARED INTENT TAXONOMY
// v0.4.2: THREAT_INTENTS + NLI_SYSTEM_PROMPT are loaded from
// core/nli-intents.json so the Python port reads identical data. Before
// v0.4.2 the JS kept inline copies that drifted whenever either language
// was tuned. The JSON is the source of truth.
// ═══════════════════════════════════════════════════════════════════════
function loadSharedIntents() {
  const p = require('path').join(__dirname, 'nli-intents.json');
  try {
    return JSON.parse(require('fs').readFileSync(p, 'utf-8'));
  } catch (e) {
    throw new Error(
      'agent-content-shield: failed to load ' + p + ': ' + e.message
    );
  }
}
const _INTENTS_FILE = loadSharedIntents();
const THREAT_INTENTS = _INTENTS_FILE.intents;
const NLI_SYSTEM_PROMPT = _INTENTS_FILE.system_prompt;

// ═══════════════════════════════════════════════════════════════════════
// Wave6-Fix: Shared JSON extraction with intent validation.
// Both Claude and Ollama paths now use this to prevent attacker-embedded
// JSON with fake "benign: true" from being accepted.
const VALID_INTENT_IDS = new Set(THREAT_INTENTS.map(t => t.id));

function extractAndValidateJson(content) {
  const jsonObjects = [];
  const jsonRx = /\{[^{}]*(?:\{[^{}]*\}[^{}]*)*\}/g;
  let jm;
  while ((jm = jsonRx.exec(content)) !== null) {
    try { jsonObjects.push(JSON.parse(jm[0])); } catch {}
  }
  // Take the last valid JSON object (model's response, not echoed input)
  const parsed = jsonObjects[jsonObjects.length - 1];
  if (!parsed) return null;
  // Validate that matched intents use known IDs — attacker can't invent fake "benign" intents
  const validMatches = (parsed.matches || []).filter(m => VALID_INTENT_IDS.has(m.intent));
  return {
    matches: validMatches,
    benign: validMatches.length === 0 ? (parsed.benign ?? true) : false,
  };
}

// ═══════════════════════════════════════════════════════════════════════
// NLI CLASSIFICATION VIA CLAUDE API
// ═══════════════════════════════════════════════════════════════════════

let anthropicClient = null;
let apiAvailable = null;

// Load API key from .env file if not in environment
function loadApiKey() {
  if (process.env.ANTHROPIC_API_KEY) return process.env.ANTHROPIC_API_KEY;
  try {
    const fs = require('fs');
    const path = require('path');
    const envPath = path.join(__dirname, '..', '.env');
    const envText = fs.readFileSync(envPath, 'utf-8');
    for (const line of envText.split('\n')) {
      const m = line.match(/^ANTHROPIC_API_KEY\s*=\s*(.+)$/);
      if (m) {
        process.env.ANTHROPIC_API_KEY = m[1].trim();
        return m[1].trim();
      }
    }
  } catch {}
  return null;
}

function getClient() {
  if (anthropicClient) return anthropicClient;
  try {
    const key = loadApiKey();
    if (!key) return null;
    const Anthropic = require('@anthropic-ai/sdk');
    anthropicClient = new Anthropic({ apiKey: key });
    return anthropicClient;
  } catch {
    return null;
  }
}


async function classifyWithClaude(text, model = 'claude-haiku-4-5-20251001') {
  const client = getClient();
  if (!client) return null;

  try {
    const response = await client.messages.create({
      model,
      max_tokens: 512,
      system: NLI_SYSTEM_PROMPT,
      // Wave5-Fix W5-04: Scan more content — attackers placed payloads past 3000 chars
      // Haiku supports 200k context, so we can safely send more
      messages: [{ role: 'user', content: `Classify this text:\n\n${text.slice(0, 8000)}` }],
    });

    const content = response.content[0]?.text || '';

    // Wave6-Fix: Validate parsed intents against known set to prevent
    // attacker-embedded JSON with fake "benign: true" from being accepted.
    const parsed = extractAndValidateJson(content);
    if (!parsed) return null;
    return { ...parsed, model, raw: content.slice(0, 300) };
  } catch (e) {
    // API error — return null to signal fallback
    return null;
  }
}

// ═══════════════════════════════════════════════════════════════════════
// FALLBACK: NLI via local Ollama (deepseek-r1:8b)
// ═══════════════════════════════════════════════════════════════════════

const OLLAMA_BASE = process.env.OLLAMA_URL || 'http://localhost:11434';

async function classifyWithOllama(text) {
  // Wave6-Fix: Removed /no_think — let the model reason for security decisions
  const prompt = `${NLI_SYSTEM_PROMPT}

Classify this text:

${text.slice(0, 2000)}`;

  try {
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), 30000);
    const res = await fetch(`${OLLAMA_BASE}/api/generate`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        model: process.env.SHIELD_CLASSIFIER_MODEL || 'deepseek-r1:8b',
        prompt,
        stream: false,
        options: { temperature: 0.0, num_predict: 512, num_ctx: 4096 },
      }),
      signal: controller.signal,
    });
    clearTimeout(timer);
    if (!res.ok) return null;

    const data = await res.json();
    const content = (data.response || '').replace(/<think>[\s\S]*?<\/think>/g, '').trim();

    // Wave6-Fix: Use same validated JSON extraction as Claude path
    const parsed = extractAndValidateJson(content);
    if (!parsed) return null;
    return { ...parsed, model: 'deepseek-r1:8b', raw: content.slice(0, 300) };
  } catch {
    return null;
  }
}

// ═══════════════════════════════════════════════════════════════════════
// MAIN NLI CLASSIFICATION FUNCTION
// ═══════════════════════════════════════════════════════════════════════

/**
 * nliClassify(text) — Classify text via NLI intent matching
 *
 * Tries Claude Haiku first (fast, high quality), falls back to
 * Claude Sonnet (if Haiku uncertain), then Ollama deepseek-r1:8b.
 *
 * Returns: {
 *   injection: boolean,
 *   confidence: number,
 *   intents: [{id, confidence, evidence}],
 *   model: string,
 *   latencyMs: number,
 * }
 */
async function nliClassify(text) {
  const start = Date.now();

  // Try Claude Haiku first (fastest, cheapest)
  let result = await classifyWithClaude(text, 'claude-haiku-4-5-20251001');

  // If Haiku returns borderline results, escalate to Sonnet
  if (result && result.matches.length > 0) {
    const maxConf = Math.max(...result.matches.map(m => m.confidence));
    if (maxConf > 0.3 && maxConf < 0.7) {
      // Borderline — escalate to Sonnet for higher quality
      const sonnetResult = await classifyWithClaude(text, 'claude-sonnet-4-6');
      if (sonnetResult) result = sonnetResult;
    }
  }

  // Fallback to Ollama if no API
  if (!result) {
    result = await classifyWithOllama(text);
  }

  // Wave4-Fix: Fail CLOSED when all classifiers are unavailable
  // Previously returned injection: false, which was the most critical bypass
  if (!result) {
    return {
      injection: true,  // Assume hostile when we can't classify
      confidence: 0.5,
      intents: [{ intent: 'classifier_unavailable', confidence: 0.5, severity: 7 }],
      maxSeverity: 7,
      model: 'none',
      latencyMs: Date.now() - start,
      warning: 'All classifiers unavailable — failing closed',
    };
  }

  // Compute aggregate injection signal
  const highConfMatches = (result.matches || []).filter(m => m.confidence >= 0.6);
  const maxConf = result.matches.length > 0
    ? Math.max(...result.matches.map(m => m.confidence || 0))
    : 0;

  // Map matched intents to severity scores
  const matchedIntents = (result.matches || []).map(m => {
    const intent = THREAT_INTENTS.find(t => t.id === m.intent);
    return {
      ...m,
      severity: intent?.severity || 7,
    };
  });

  return {
    injection: highConfMatches.length > 0,
    confidence: maxConf,
    intents: matchedIntents,
    maxSeverity: matchedIntents.length > 0 ? Math.max(...matchedIntents.map(m => m.severity)) : 0,
    model: result.model,
    benign: result.benign,
    latencyMs: Date.now() - start,
  };
}

// ═══════════════════════════════════════════════════════════════════════
// EXPORTS
// ═══════════════════════════════════════════════════════════════════════

module.exports = {
  nliClassify,
  classifyWithClaude,
  classifyWithOllama,
  THREAT_INTENTS,
  NLI_SYSTEM_PROMPT,
};

File: core/nli_classifier.py
"""
Agent Content Shield — NLI Intent Classifier (Python)

Parity port of core/nli-classifier.js. Same fail-closed behavior: if
all classifier paths fail, return ``injection=True`` with a warning
rather than silently passing — an unreachable classifier is the worst
possible moment to fail open.

Zero *required* dependencies. Uses stdlib ``urllib`` for both the
Anthropic API path and the Ollama fallback. Callers that already have
``anthropic`` installed can pass ``use_sdk=True`` for connection
pooling + retries; otherwise the stdlib path works identically.

Order of attempts:
  1. Anthropic API via Claude Haiku (fastest)
  2. Anthropic API via Claude Sonnet if Haiku result is borderline
  3. Ollama deepseek-r1:8b fallback
  4. Fail closed (inject=True) if all paths fail

Shared state: core/nli-intents.json holds the 8-item threat taxonomy
and the NLI system prompt. JS + Python load the same file (JS still
uses inline copies in v0.4.1 — v0.4.2 tracks the JS refactor).
"""

from __future__ import annotations

import json
import os
import re
import time
import urllib.error
import urllib.request
from pathlib import Path
from typing import Any

# ── Shared lexicon ───────────────────────────────────────────────


def _intents_path() -> Path:
    return Path(__file__).resolve().parent / "nli-intents.json"


def _load_intents() -> dict[str, Any]:
    try:
        return json.loads(_intents_path().read_text(encoding="utf-8"))
    except FileNotFoundError as e:
        raise RuntimeError(
            f"nli-intents.json missing at {_intents_path()}; v0.4.1 requires "
            "the shared intent taxonomy to be present"
        ) from e


_INTENTS_FILE = _load_intents()
THREAT_INTENTS: list[dict[str, Any]] = _INTENTS_FILE["intents"]
NLI_SYSTEM_PROMPT: str = _INTENTS_FILE["system_prompt"]
_VALID_INTENT_IDS: set[str] = {t["id"] for t in THREAT_INTENTS}


# ── JSON extraction + validation ─────────────────────────────────

_JSON_RX = re.compile(r"\{[^{}]*(?:\{[^{}]*\}[^{}]*)*\}", re.DOTALL)


def extract_and_validate_json(content: str) -> dict[str, Any] | None:
    """Pull the LAST JSON object from the classifier's raw response
    (last = the model's reply, not an echoed input). Filter out any
    matches with an intent id not in our taxonomy — an attacker-
    embedded JSON with a fabricated `{"benign": true}` must not be
    able to override the classifier's actual decision."""
    parsed_objects: list[dict[str, Any]] = []
    for match in _JSON_RX.finditer(content):
        try:
            parsed_objects.append(json.loads(match.group(0)))
        except json.JSONDecodeError:
            continue
    if not parsed_objects:
        return None
    parsed = parsed_objects[-1]
    valid_matches = [
        m for m in (parsed.get("matches") or []) if m.get("intent") in _VALID_INTENT_IDS
    ]
    benign_in_payload = parsed.get("benign")
    return {
        "matches": valid_matches,
        "benign": (benign_in_payload if benign_in_payload is not None else True)
        if not valid_matches
        else False,
    }


# ── Anthropic API client (stdlib urllib, no SDK required) ───────


def _load_api_key() -> str | None:
    if os.environ.get("ANTHROPIC_API_KEY"):
        return os.environ["ANTHROPIC_API_KEY"]
    # Look for a sibling .env file. We never walk the filesystem — the
    # shield config format is always next to the module.
    env_path = Path(__file__).resolve().parent.parent / ".env"
    try:
        for line in env_path.read_text(encoding="utf-8").splitlines():
            m = re.match(r"^ANTHROPIC_API_KEY\s*=\s*(.+)$", line)
            if m:
                value = m.group(1).strip().strip('"').strip("'")
                os.environ["ANTHROPIC_API_KEY"] = value
                return value
    except OSError:
        pass
    return None


def classify_with_claude(
    text: str, *, model: str = "claude-haiku-4-5-20251001", timeout_sec: float = 15.0
) -> dict[str, Any] | None:
    """Call the Claude Messages API directly via urllib. Returns
    validated-intents dict or None on any failure. Never raises."""
    api_key = _load_api_key()
    if not api_key:
        return None
    body = json.dumps(
        {
            "model": model,
            "max_tokens": 512,
            "system": NLI_SYSTEM_PROMPT,
            "messages": [{"role": "user", "content": f"Classify this text:\n\n{text[:8000]}"}],
        }
    ).encode("utf-8")
    req = urllib.request.Request(
        "https://api.anthropic.com/v1/messages",
        data=body,
        headers={
            "content-type": "application/json",
            "x-api-key": api_key,
            "anthropic-version": "2023-06-01",
        },
        method="POST",
    )
    try:
        with urllib.request.urlopen(req, timeout=timeout_sec) as resp:
            if resp.status != 200:
                return None
            data = json.loads(resp.read().decode("utf-8"))
    except (urllib.error.URLError, TimeoutError, json.JSONDecodeError):
        return None
    except Exception:
        return None

    content_blocks = data.get("content", [])
    text_content = ""
    for block in content_blocks:
        if block.get("type") == "text":
            text_content += block.get("text", "")
    if not text_content:
        return None
    parsed = extract_and_validate_json(text_content)
    if not parsed:
        return None
    return {**parsed, "model": model, "raw": text_content[:300]}


# ── Ollama fallback ─────────────────────────────────────────────

OLLAMA_BASE = os.environ.get("OLLAMA_URL", "http://localhost:11434")


def classify_with_ollama(text: str, *, timeout_sec: float = 30.0) -> dict[str, Any] | None:
    """Fallback via local Ollama deepseek-r1:8b (or another configured
    classifier model). Only used when the Anthropic path is unavailable."""
    model = os.environ.get("SHIELD_CLASSIFIER_MODEL", "deepseek-r1:8b")
    prompt = f"{NLI_SYSTEM_PROMPT}\n\nClassify this text:\n\n{text[:2000]}"
    body = json.dumps(
        {
            "model": model,
            "prompt": prompt,
            "stream": False,
            "options": {"temperature": 0.0, "num_predict": 512, "num_ctx": 4096},
        }
    ).encode("utf-8")
    req = urllib.request.Request(
        f"{OLLAMA_BASE}/api/generate",
        data=body,
        headers={"Content-Type": "application/json"},
        method="POST",
    )
    try:
        with urllib.request.urlopen(req, timeout=timeout_sec) as resp:
            if resp.status != 200:
                return None
            data = json.loads(resp.read().decode("utf-8"))
    except Exception:
        return None

    raw = data.get("response", "") or ""
    # Strip <think>...</think> reasoning blocks the deepseek-r1 family emits.
    cleaned = re.sub(r"<think>[\s\S]*?</think>", "", raw).strip()
    parsed = extract_and_validate_json(cleaned)
    if not parsed:
        return None
    return {**parsed, "model": model, "raw": cleaned[:300]}


# ── Main entrypoint ──────────────────────────────────────────────


def nli_classify(text: str) -> dict[str, Any]:
    """Classify text for injection intent. Tries Haiku → Sonnet (if
    borderline) → Ollama fallback. Fails CLOSED — returns
    ``injection=True`` with a warning when all classifiers fail."""
    start = time.time()
    result = classify_with_claude(text, model="claude-haiku-4-5-20251001")

    if result and result.get("matches"):
        confidences = [m.get("confidence", 0) for m in result["matches"]]
        max_conf = max(confidences) if confidences else 0
        if 0.3 < max_conf < 0.7:
            # Borderline — escalate to Sonnet for a better verdict
            sonnet = classify_with_claude(text, model="claude-sonnet-4-6")
            if sonnet:
                result = sonnet

    if not result:
        result = classify_with_ollama(text)

    if not result:
        # Fail closed — an unreachable classifier is the worst moment
        # to assume text is benign.
        return {
            "injection": True,
            "confidence": 0.5,
            "intents": [
                {"intent": "classifier_unavailable", "confidence": 0.5, "severity": 7}
            ],
            "maxSeverity": 7,
            "model": "none",
            "latencyMs": int((time.time() - start) * 1000),
            "warning": "All classifiers unavailable — failing closed",
        }

    matches = result.get("matches") or []
    confidences = [m.get("confidence", 0) for m in matches]
    high_conf = [m for m in matches if m.get("confidence", 0) >= 0.6]
    max_conf = max(confidences) if confidences else 0

    matched_intents = []
    for m in matches:
        intent = next((t for t in THREAT_INTENTS if t["id"] == m.get("intent")), None)
        matched_intents.append({**m, "severity": intent["severity"] if intent else 7})
    max_severity = max(
        (m.get("severity", 0) for m in matched_intents), default=0
    )

    return {
        "injection": len(high_conf) > 0,
        "confidence": max_conf,
        "intents": matched_intents,
        "maxSeverity": max_severity,
        "model": result.get("model"),
        "benign": result.get("benign"),
        "latencyMs": int((time.time() - start) * 1000),
    }


__all__ = [
    "nli_classify",
    "classify_with_claude",
    "classify_with_ollama",
    "extract_and_validate_json",
    "THREAT_INTENTS",
    "NLI_SYSTEM_PROMPT",
]
